lib/platform/git/index.js

```diff
@@ -0,0 +1,169 @@
+    // not supported yet
+    await git.raw(['config', 'commit.gpgsign', 'false']);
+  }
+
+  config.git = git;
+  // see https://stackoverflow.com/a/44750379/1438522
+  config.baseBranch = (await git.raw([
+    'symbolic-ref',
+    'refs/remotes/origin/HEAD',
+  ]))
+    .replace('refs/remotes/origin/', '')
+    .trim();
+  config.dir = tmpDir;
+}
+
+function setBaseBranch(branchName) {
+  if (branchName) {
+    logger.debug(`Setting baseBranch to ${branchName}`);
+    config.baseBranch = branchName;
+  }
+}
+
+async function getFileList(branchName = config.baseBranch) {
+  const exists = await branchExists(branchName);
+  if (!exists) {
+    return [];
+  }
+  const files = await config.git.raw([
+    'ls-tree',
+    '-r',
+    '--name-only',
+    'origin/' + branchName,
+  ]);
+  return files.split('\n').filter(Boolean);
+}
+
+function localName(branchName) {
+  return branchName.replace(/^origin\//, '');
+}
+
+async function branchExists(branchName) {
+  const branches = await config.git.branch(['--remotes', '--verbose']);
+  return branches.all.map(localName).includes(branchName);
+}
+
+async function getAllRenovateBranches(branchPrefix) {
+  const branches = await config.git.branch(['--remotes', '--verbose']);
+  return branches.all
+    .map(localName)
+    .filter(branchName => branchName.startsWith(branchPrefix));
+}
+
+async function isBranchStale(branchName) {
+  const branches = await config.git.branch([
+    '--remotes',
+    '--verbose',
+    '--contains',
+    config.baseBranch,
+  ]);
+  return !branches.all.map(localName).includes(branchName);
+}
+
+async function deleteBranch(branchName) {
+  await config.git.raw(['push', '--delete', 'origin', branchName]);
+  try {
+    await config.git.deleteLocalBranch(branchName);
+  } catch (ex) {
+    // local branch may not exist
+  }
+}
+
+async function mergeBranch(branchName) {
+  await config.git.reset('hard');
+  await config.git.checkoutBranch(branchName, 'origin/' + branchName);
+  await config.git.checkout(config.baseBranch);
+  await config.git.merge([branchName]);
+  await config.git.push('origin', config.baseBranch);
+}
+
+async function getBranchLastCommitTime(branchName) {
+  try {
+    const time = await config.git.show(['-s', '--format=%ai', branchName]);
+    return new Date(Date.parse(time));
+  } catch (ex) {
+    return new Date();
+  }
+}
+
+async function getFile(filePath, branchName) {
+  if (branchName) {
+    const exists = await branchExists(branchName);
+    if (!exists) {
+      return null;
+    }
+  }
+  try {
+    const content = await config.git.show([
+      (branchName || config.baseBranch) + ':' + filePath,
+    ]);
+    return content;
+  } catch (ex) {
+    return null;
+  }
+}
+
+async function commitFilesToBranch(
+  branchName,
+  files,
+  message,
+  parentBranch = config.baseBranch
+) {
+  await config.git.reset('hard');
+  const exists = await branchExists(branchName);
+  if (exists) {
+    await config.git.checkout(branchName);
```

Comment:
TODO: We need to check if this is how it works currently. I think we maybe "force push" over the top of any existing branch even if it exists, because we always want just 1 commit per branch